control/prometheus.py

```diff
@@ -0,0 +1,308 @@
+#
+#  Copyright (c) 2021 International Business Machines
+#  All rights reserved.
+#
+#  SPDX-License-Identifier: LGPL-3.0-or-later
+#
+#  Authors: pcuzner@ibm.com
+#
+
+import os
+import time
+import logging
+import threading
+import spdk.rpc as rpc
+
+from prometheus_client.core import REGISTRY, GaugeMetricFamily, CounterMetricFamily, InfoMetricFamily
+from prometheus_client import start_http_server, GC_COLLECTOR
+from typing import NamedTuple
+
+COLLECTION_ELAPSED_WARNING = 0.8   # Percentage of the refresh interval before a warning message is issued
+REGISTRY.unregister(GC_COLLECTOR)  # Turn of the garbage collector metrics
+
+
+class RBD(NamedTuple):
+    pool: str
+    namespace: str
+    image: str
+
+
+def timer(method):
+    def call(self, *args, **kwargs):
+        st = time.time()
+        result = method(self, *args, **kwargs)
+        elapsed = time.time() - st
+        if hasattr(self, 'method_timings'):
+            self.method_timings[method.__name__] = elapsed
+        return result
+    return call
+
+
+def start_httpd(**kwargs):
+    """Start the prometheus http endpoint, catching any exception"""
+    logger = logging.getLogger(__name__)
+    try:
+        start_http_server(**kwargs)
+    except Exception as e:
+        logger.error("Failed to start the prometheus http server", exc_info=True)
+        return False
+    return True
+
+
+def start_exporter(spdk_rpc_client, config):
+    """Start the prometheus exporter and register the NVMeOF custom collector"""
+    logger = logging.getLogger(__name__)
+
+    port = config.getint_with_default("gateway", "prometheus_port", 10008)
+    ssl = config.getboolean_with_default("gateway", "prometheus_exporter_ssl", True)
+    mode = 'https' if ssl else 'http'
+
+    if ssl:
+        cert_filepath = config.get('mtls', 'server_cert')
+        key_filepath = config.get('mtls', 'server_key')
+
+        if os.path.exists(cert_filepath) and os.path.exists(key_filepath):
+            httpd_ok = start_httpd(port=port, certfile=cert_filepath, keyfile=key_filepath)
+        else:
+            httpd_ok = False    
+            logger.error("Unable to start prometheus exporter - missing cert/key file(s)")
+    else:
+        # SSL mode explicitly disabled by config option
+        httpd_ok = start_httpd(port=port)
+    
+    if httpd_ok:
+        logger.info(f"Prometheus exporter running in {mode} mode, listening on port {port}")
+        REGISTRY.register(NVMeOFCollector(spdk_rpc_client, config))
```

Comment:
Nit: this kind of flows is where exceptions work best:
```suggestion
    try:
      if ssl:         
          # No need the check for certificates, this method already raises an Exception if not found
          start_http_server(
              port=port,
              certfile=config.get('mtls', 'server_cert'),
              keyfile=config.get('mtls', 'server_key')
              )
      else:
         # SSL mode explicitly disabled by config option
          httpd_ok = start_httpd(port=port)
    except Exception:
        # logger.exception() automatically attaches the Exception raised
        logger.exception("Failed to start the prometheus http server")
    else:
        logger.info(f"Prometheus exporter running in {mode} mode, listening on port {port}")
        REGISTRY.register(NVMeOFCollector(spdk_rpc_client, config))
```

Reply:
I check for files to give more meaningful error messages.